presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java

```
@@ -514,22 +516,85 @@ public synchronized MetastoreOperationResult renameTable(MetastoreContext metast
+        Path newMetadataDirectory = getTableMetadataDirectory(newDatabaseName, newTableName);
+
+        if (isIcebergTable(table)) {
+            renameIcebergTable(metadataDirectory, newMetadataDirectory);
+        }
+        else {
+            renameTable(metadataDirectory, newMetadataDirectory);
+        }
+
+        return EMPTY_RESULT;
+    }
+
+    private void renameIcebergTable(Path originalMetadataDirectory, Path newMetadataDirectory)
+    {
+        Optional<Runnable> rollbackAction = Optional.empty();
         try {
-            if (!metadataFileSystem.rename(getTableMetadataDirectory(databaseName, tableName), getTableMetadataDirectory(newDatabaseName, newTableName))) {
-                throw new PrestoException(HIVE_METASTORE_ERROR, "Could not rename table directory");
+            // If the directory `.prestoPermissions` exists, copy it to the new table metadata directory
+            Path originTablePermissionDir = new Path(originalMetadataDirectory, PRESTO_PERMISSIONS_DIRECTORY_NAME);
+            Path newTablePermissionDir = new Path(newMetadataDirectory, PRESTO_PERMISSIONS_DIRECTORY_NAME);
+            if (metadataFileSystem.exists(originTablePermissionDir)) {
+                if (!FileUtil.copy(metadataFileSystem, originTablePermissionDir,
+                        metadataFileSystem, newTablePermissionDir, false, metadataFileSystem.getConf())) {
+                    throw new IOException(format("Could not rename table. Failed to copy directory: %s to %s", originTablePermissionDir, newTablePermissionDir));
+                }
+                else {
+                    rollbackAction = Optional.of(() -> {
+                        try {
+                            metadataFileSystem.delete(newTablePermissionDir, true);
+                        }
+                        catch (IOException e) {
+                            LOG.warn("Could not delete table permission directory: %s", newTablePermissionDir);
```

Same here, should we ignore exception here?

---

The same as above.